zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/LearnerMaster.java

```diff
@@ -29,18 +31,81 @@
 /**
  * interface for keeping Observers in sync
  */
-public interface LearnerMaster {
+public abstract class LearnerMaster {
+    private static final Logger LOG = LoggerFactory.getLogger(LearnerMaster.class);
+
+    // Throttle when there are too many concurrent snapshots being sent to observers
+    private static final String MAX_CONCURRENT_SNAPSYNCS = "zookeeper.leader.maxConcurrentSnapSyncs";
```

Comment:
Would this configuration be shared between leader and observer master? If so, I think there are a few assumptions that worth a closer look.

1) Leader and Followers are running on machines with similar capability (CPU, network bandwidth). This assumption is usually true.
2) ObserverMasters and Observers are also running on machines with similar capabilities (CPU, network bandwidth). This assumption may not be true. E.g. If we deploy some Observers in a different data center than the ObserverMaster, then the effective network bandwidth between ObserverMaster and Observer could be significantly smaller than the bandwidth between Leader and Followers. In this case, we might want to set a different maxConcurrentSnapSyncs for ObserverMaster than Leader.

Does those assumptions make sense here?

Reply:
The thing is, we don't know what the appropriate thresholds are until we roll the feature out and then tune the settings.  If it turns out that we do need different settings for Leader and ObserverMaster, we can add it, instead of introducing a setting that might never be used. What do you think? (BTW, in our deployment, Leader and Followers can be in different data centers too.)

Reply:
Thanks for the details! I think we can keep this configuration just one parameter for now. I'd love to learn the actual system behavior once this is used in production.